### ISSUE TITLE
No such variable t

### ISSUE THREAD

**Issue by nicolabotta** [idris-lang/Idris-dev#3654]
I am likely missing something obvious but I do not understand why the code
```
> %default total
> %access public export
> %auto_implicits off

> Natural : {F, G : Type -> Type} -> 
>           (Functor F) => (Functor G) =>
>           (t : {A : Type} -> F A -> G A) -> 
>           Type            
> Natural {F} {G} t = {A, B : Type} -> 
>                     (f : A -> B) ->
>                     (x : F A) -> 
>                     t (map f x) = map f  (t x) 

> Monotone : {B, C : Type} -> {F : Type -> Type} -> (Functor F) => 
>            (LTE_B : B -> B -> Type) -> 
>            (LTE_C : C -> C -> Type) -> 
>            (measure : F B -> C) -> 
>            Type
> Monotone {B} {C} {F} LTE_B LTE_C measure =
>   {A : Type} ->
>   (f : A -> B) -> 
>   (g : A -> B) -> 
>   (p : (a : A) -> f a `LTE_B` g a) -> 
>   (x : F A) -> 
>   measure (map f x) `LTE_C` measure (map g x)  

> monotoneNaturalLemma: {B, C : Type} -> {F : Type -> Type} -> (Functor F) => 
>                       (LTE_B : B -> B -> Type) -> 
>                       (LTE_C : C -> C -> Type) -> 
>                       (measure : F B -> C) ->
>                       Monotone LTE_B LTE_C measure -> 
>                       (t : {A : Type} -> F A -> F A) -> 
>                       Natural t -> 
>                       Monotone LTE_B LTE_C (measure . t)
>
> monotoneNaturalLemma {B} {C} {F} LTE_B LTE_C m mm t nt = mmt where
>   mmt : {A : Type} -> 
>         (f : A -> B) -> 
>         (g : A -> B) -> 
>         (p : (a : A) -> f a `LTE_B` g a) ->
>         (x : F A) -> 
>         m (t {A = B} (map f x)) `LTE_C` m (t {A = B} (map g x))   
>   mmt = ?kika
```
fails to type check with
```
- + Errors (1)
 `-- ./Functor/issues/no_such_variable.lidr line 37 col 8:
     When checking type of Main.monotoneNaturalLemma, mmt:
     When checking an application of LTE_C:
             No such variable t
```
Is this a blind spot or an Idris bug?

**Comment by ahmadsalim** [comment]
This might be a bug with implicits in higher-order function arguments.
As far as I remember "plicity" is only stored for top-level functions (perhaps this should change).

In any case, thanks for the bug-report.

**Comment by ahmadsalim** [comment]
Have you tried using an explicit function instead, and do you experience the same error?

**Comment by nicolabotta** [comment]
@ahmadsalim I am not sure what you mean precisely by "using an explicit function". I have initially tried
```
> monotoneNaturalLemma {B} {C} {F} LTE_B LTE_C m mm t nt = mmt where
>   mmt : Monotone {B} {C} {F} LTE_B LTE_C (m . t)
>   mmt = ?kika
```
instead of the above but this yields
```
- + Errors (1)
 `-- (val) line 0 col -1:
     When checking type of Main.monotoneNaturalLemma, mmt:
     When checking an application of t:
             Type mismatch between
                     Type (Type of A1)
             and
                     F A1 (Expected type)
```
which also suggests that there is something wrong in Idris, at least in the generation of the error message.

**Comment by ahmadsalim** [comment]
@nicolabotta I meant changing the type of `t` from `{A : Type} -> F A -> F A` to `(A : Type) -> F A -> F A`

**Comment by nicolabotta** [comment]
@ahmadsalim I'm pretty sure that that would work. But it would make the notion of naturality
```
> Natural' : {F, G : Type -> Type} -> 
>            (Functor F) => (Functor G) =>
>            (t : (A : Type) -> F A -> G A) -> 
>            Type            
> Natural' {F} {G} t = {A, B : Type} -> 
>                      (f : A -> B) ->
>                      (x : F A) -> 
>                      t B (map f x) = map f (t A x)
```
a bit unnatural.

**Comment by nicolabotta** [comment]
Strangely enough, a `let`based implementation of the lemma type checks without problems:
```
> monotoneNaturalLemma LTE_B LTE_C m mm t nt = mmt where
>   mmt f g p x = let s1 = mm f g p (t x) in
>                 let s2 = replace {P = \ X => m X `LTE_C` m (map g (t x))} (sym (nt f x)) s1 in
>                 let s3 = replace {P = \ X => m (t (map f x)) `LTE_C` m X} (sym (nt g x)) s2 in
>                 s3
```
whereas even
```
> monotoneNaturalLemma {B} {C} {F} LTE_B LTE_C m mm t nt = mmt where
>   mmt : Monotone LTE_B LTE_C (m . t)
>   mmt = ?kika
```
fails to type check!

**Comment by edwinb** [comment]
I've now looked into this, as part of my current attempt to polish up scoped implicits so that they work properly. The issue is that the `t` is given the wrong type when passed to the where block, with an explicit argument given instead of an implicit one.

The reason for the 'No such variable' error is that, for some reason, that's the error I chose to give when there's too many arguments given to a locally defined function, rather than the more sensible "Too many arguments".

Anyway. Now that I can see what's wrong it'll help me to get the implicits working properly in general...
